packages/edit-navigation/src/components/menu-editor/use-navigation-blocks.js

```diff
@@ -41,46 +46,76 @@ export default function useNavigationBlocks( menuId ) {
 			return;
 		}
 
-		menuItemsRef.current = {};
+		const itemIndex = {};
+		menuItems.forEach( ( item ) => {
+			if ( ! itemIndex[ item.parent ] ) {
+				itemIndex[ item.parent ] = [];
+			}
+			itemIndex[ item.parent ].push( item );
+		} );
 
-		const innerBlocks = [];
+		menuItemsRef.current = {};
 
-		for ( const menuItem of menuItems ) {
-			const block = createBlockFromMenuItem( menuItem );
-			menuItemsRef.current[ block.clientId ] = menuItem;
-			innerBlocks.push( block );
-		}
+		const createMenuItemBlocks = ( items ) => {
+			const innerBlocks = [];
+			for ( const item of items ) {
+				let menuItemInnerBlocks = [];
+				if ( itemIndex[ item.id ] && itemIndex[ item.id ].length > 0 ) {
+					menuItemInnerBlocks = createMenuItemBlocks(
+						itemIndex[ item.id ]
+					);
+				}
+				const block = createBlockFromMenuItem(
+					item,
+					menuItemInnerBlocks
+				);
+				menuItemsRef.current[ block.clientId ] = item;
+				innerBlocks.push( block );
+			}
+			return innerBlocks;
+		};
 
+		const innerBlocks = createMenuItemBlocks( itemIndex[ 0 ] );
 		setBlocks( [ createBlock( 'core/navigation', {}, innerBlocks ) ] );
 	}, [ menuItems ] );
 
 	const saveBlocks = () => {
-		const { innerBlocks } = blocks[ 0 ];
-
-		for ( const block of innerBlocks ) {
-			const menuItem = menuItemsRef.current[ block.clientId ];
-
-			if ( ! menuItem ) {
-				saveMenuItem( {
-					...createMenuItemAttributesFromBlock( block ),
-					menus: menuId,
-				} );
-				continue;
+		const { clientId, innerBlocks } = blocks[ 0 ];
+
+		const saveNestedBlocks = ( nestedBlocks, parentId ) => {
+			for ( const block of nestedBlocks ) {
+				if ( block.innerBlocks && block.innerBlocks.length > 0 ) {
+					saveNestedBlocks( block.innerBlocks, block.clientId );
+				}
+				const menuItem = menuItemsRef.current[ block.clientId ];
+				const parentItemId =
+					menuItemsRef.current[ parentId ] &&
+					menuItemsRef.current[ parentId ].id;
+				if ( ! menuItem ) {
+					saveMenuItem( {
+						...createMenuItemAttributesFromBlock( block ),
+						menus: menuId,
+						parent: parentItemId || 0,
+					} );
+					continue;
+				}
+
+				if (
+					! isEqual(
+						block.attributes,
+						createBlockFromMenuItem( menuItem ).attributes
+					)
+				) {
+					saveMenuItem( {
+						...menuItem,
+						...createMenuItemAttributesFromBlock( block ),
+						menus: menuId, // Gotta do this because REST API doesn't like receiving an array here. Maybe a bug in the REST API?
```

Comment:
Do we need to set `parent` here? What happens if a Navigation Link is moved from one submenu to another?

Reply:
Good question; it doesn't appear to be possible to move items between nesting levels at the moment. Also, I just realised that nav items can be _dragged_ by the handle from the top toolbar and the experience is... not delightful 🙀

Reply:
We should still send parent even if our UI is missing proper nesting, I think!